Implement CSS Grid

nicoburns opened the pull request:
# Objective

Fixes #204

## Feedback wanted

Nothing in particular just yet. This PR is still very incomplete. Just putting it up early to track progress.

TimJentzsch commented:
FYI, if your work is still in progress it's good practice to mark the PR as draft: [GitHub Docs](https://github.blog/2019-02-14-introducing-draft-pull-requests/)

This prevents merging of the PR before it's ready.

You should be able to mark it as draft retroactively in the sidebar on the right, if IIRC.

Weibye commented on docs/style-properties.md:
```diff
@@ -0,0 +1,55 @@
+
+
+## Style Properties
```
This is a really good addition!

Weibye commented on src/style.rs:
```diff
@@ -362,34 +477,59 @@ impl Default for Size<Dimension> {
+
+    // Gap
+    /// How large should the gaps between items in a grid or flex container be?
+    pub gap: Size<Dimension>,
```
We should try to get the flexbox gap implemented shortly after this or alongside of it

nicoburns replied:
Agreed. Flexbox implementation for `justify_items` and `justify_self` is also missing (see: https://github.com/DioxusLabs/taffy/issues/204#issuecomment-1193202234)

Weibye commented on src/style.rs:
```diff
@@ -398,19 +538,28 @@ pub struct FlexboxLayout {
-    ///
-    /// The ratio is calculated as width divided by height.
-    pub aspect_ratio: Option<f32>,
+
+    // Grid container properies
+    /// Defines the track sizing functions (widths) of the grid rows
+    pub grid_template_rows: GridTrackVec<GridTrackSizingFunction>,
+    /// Defines the track sizing functions (heights) of the grid columns
+    pub grid_template_columns: GridTrackVec<GridTrackSizingFunction>,
+    /// Defines the size of implicitly created rows
+    pub grid_auto_rows: GridTrackVec<GridTrackSizingFunction>,
+    /// Defined the size of implicitly created columns
+    pub grid_auto_columns: GridTrackVec<GridTrackSizingFunction>,
+    /// Controls how items get placed into the grid for auto-placed items
+    pub grid_auto_flow: GridAutoFlow,
+
+    // Grid child properties
+    /// Defines which row in the grid the item should start at
+    pub grid_row_start: GridLine,
+    /// Defines which row in the grid the item should end at
+    pub grid_row_end: GridLine,
+    /// Defines which column in the grid the item should start at
+    pub grid_column_start: GridLine,
+    /// Defines which column in the grid the item should end at
+    pub grid_column_end: GridLine,
```
Should this be split out into a separate type? 

Would it make sense to have a `FlexboxLayout` and a different `GridLayout`?

nicoburns replied:
If we really want to split it out, I imagine something like below would make sense. I worry that adding static typing constraints just makes thing unwieldy though. Especially as I imagine it will be rare for Taffy to be used directly by hand, and much more common for it to be translated from the surface syntax of another library/system. I wonder if it might not be better to be permission with what we accept, but print diagnostic information about ignored styles (or return it to the caller in a machine inspectable format).

```rust
enum InnerDisplay {
    Flex(FlexContainerStyle),
    Grid(GridContainerStyle),
    Leaf,
}

struct FlexContainerStyle {
    flex_direction: FlexDirection,
    flex_wrap: FlexWrap,

    align_content: AlignContent,
    align_items: AlignItems,
    justify_content: JustifyContent,
    justify_items: JustifyItems,
    gap: Size<Dimension>,
}


struct GridContainerStyle {
    grid_template_rows: GridTrackVec<GridTrackSizingFunction>,
    grid_template_columns: GridTrackVec<GridTrackSizingFunction>,
    grid_auto_rows: GridTrackVec<GridTrackSizingFunction>,
    grid_auto_columns: GridTrackVec<GridTrackSizingFunction>,
    grid_auto_flow: GridAutoFlow,

    align_content: AlignContent,
    align_items: AlignItems,
    justify_content: JustifyContent,
    justify_items: JustifyItems,
    gap: Size<Dimension>,
}

enum OuterDisplay {
    None,
    Root,
    FlexChild(FlexChildStyle),
    GridChild(GridChildStyle),
}

struct FlexChildStyle {
    flex_basis: Dimension,
    flex_grow: f32,
    flex_shrink: f32,
}

struct GridChildStyle {
    grid_row_start: GridLine,
    grid_row_end: GridLine,
    grid_column_start: GridLine,
    grid_column_end: GridLine,
}

struct CssStyle {
    /// What layout strategy should be used?
    outer_display: OuterDisplay,
    inner_display: InnerDisplay,

    // Position properties
    position_type: PositionType,
    position: Rect<Dimension>,

    // Size properies
    size: Size<Dimension>,
    min_size: Size<Dimension>,
    max_size: Size<Dimension>,
    aspect_ratio: Option<f32>,
    margin: Rect<Dimension>,
    padding: Rect<Dimension>,
    border: Rect<Dimension>,
}
```

alice-i-cecile replied:
This code base is hard enough to maintain as it is; I'm alright to increase the static typing to make it harder for us to have bugs.

Like you say, most end users will be using taffy in a wrapped form.

Weibye replied:
I think this design approach can make sense and in the direction I was picturing. 
It might be that we should wait until we have a 3rd layouting algorithm to make sure that what we build is general across the correct domain.

@alice-i-cecile any opinion?

alice-i-cecile replied:
This is the direction I was picturing too. I'd like to avoid adding `InnerDisplay` and `OuterDisplay` for this PR; it's a lot of complexity and can be nicely severed.

nicoburns replied:
I now think we ought to do something similar to https://github.com/bevyengine/bevy/pull/5513, but I agree that it ought to be a separate PR (probably landing after this one, as it'll be easier to work out the correct splits for the styles once we have all of them).

alice-i-cecile commented:
As a heads up @nicoburns, I'd like to merge this relatively aggressively, then iterating on it in follow-up PRs before the next major version. Great work so far!

It's a much easier workflow for collaboration than trying to get it perfect in the very first PR.

nicoburns commented:
@alice-i-cecile Good to know! I can break this up into smaller PRs if that would be helpful? And maybe add feature flags for flexbox and grid (flexbox defaulting to on, grid to off for now)?

However this PR builds on top of https://github.com/DioxusLabs/taffy/pull/202 (as it already much of the work of nicely factoring out the existing flexbox implementation into it's own module). I'm not sure what the status on that PR is, but currently this one would be blocked on that one being merged.

alice-i-cecile commented:
Awesome! I'm wildly impressed with your work here.

What type of review would be most useful to you?

alice-i-cecile commented on src/axis.rs:
```diff
@@ -0,0 +1,85 @@
+//! Axis enums representing CSS Grid axis
```
Should we move the flexbox axes (i.e. FlexDirection) into this file too?

nicoburns replied:
We could do. I've been trying to keep the `flex` and `grid` stuff separate in case we want to split it into separate crates and/or feature flag flexbox as well as grid. Currently everything in `axis` is CSS-specific, but not specific to any particular layout algorithm (CSS Grid doesn't have any non-standard axes).

alice-i-cecile commented on docs/style-properties.md:
```diff
@@ -0,0 +1,52 @@
+# Style Properties
```
I love this file. Should we update what's implemented?

alice-i-cecile reviewed (changes requested):
This needs release notes, ideally with a migration guide for the small changes needed to existing code.

alice-i-cecile commented on src/style/dimension.rs:
```diff
@@ -0,0 +1,306 @@
+/// This is commonly combined with [`Rect`], [`Point`](crate::geometry::Point) and [`Size<T>`].
+#[derive(Copy, Clone, PartialEq, Debug)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub enum Dimension {
```
This doesn't seem distinct from `LengthPercentageAuto`? This name is way nicer though!

nicoburns replied:
It's not currently, but properties using `Dimension` support additional variants like `MinContent`, `MaxContent`, and `FitContent` that we don't support yet (but will probably want to at some point). As of the next release I wouldn't expect people to be using these types directly in any case. There was discussion of this on https://github.com/DioxusLabs/taffy/pull/269 where these types were introduced.

alice-i-cecile commented on src/style/grid.rs:
```diff
@@ -0,0 +1,446 @@
+use core::cmp::{max, min};
+
+/// Controls whether grid items are placed row-wise or column-wise. And whether the sparse or dense packing algorithm is used.
+/// The "dense" packing algorithm attempts to fill in holes earlier in the grid, if smaller items come up later. This may cause items to appear out-of-order, when doing so would fill in holes left by larger items.
```
```suggestion
///
/// The "dense" packing algorithm attempts to fill in holes earlier in the grid, if smaller items come up later. This may cause items to appear out-of-order, when doing so would fill in holes left by larger items.
```

Line breaks will lead to nicer tooltips in IDEs.

alice-i-cecile commented on src/style/grid.rs:
```diff
@@ -0,0 +1,446 @@
+    }
+}
+
+/// A track placement specicification. Used for grid-[row/column]-[start/end]. Named tracks are not implemented.
```
```suggestion
/// A track placement specification. Used for grid-[row/column]-[start/end]. Named tracks are not implemented.
```

alice-i-cecile commented on src/style/grid.rs:
```diff
@@ -0,0 +1,446 @@
+}
+
+impl GridPlacement {
+    /// Apply a mapping function if the GridPlacement is a Track. Otherwise return `self` unmodified.
```
```suggestion
    /// Apply a mapping function if the [`GridPlacement`] is a `Track`. Otherwise return `self` unmodified.
```

alice-i-cecile commented on src/style/grid.rs:
```diff
@@ -0,0 +1,446 @@
+    }
+}
+
+/// Represents the start and end points of a GridItem within a given axis
```
Does this docstring belong here? It needs more context if so.

alice-i-cecile commented on src/style/grid.rs:
```diff
@@ -0,0 +1,446 @@
+    }
+}
+
+/// Minimum track sizing function
```
This comment both seems wrong and unhelpful.

alice-i-cecile commented on src/style/grid.rs:
```diff
@@ -0,0 +1,446 @@
+    }
+}
+
+/// Minimum track sizing function
```
Ditto here on this description being unhelpful.

Also, line breaks :)

alice-i-cecile commented on src/style/grid.rs:
```diff
@@ -0,0 +1,446 @@
+impl MinTrackSizingFunction {
+    /// Returns fixed point values directly. Attempts to resolve percentage values against
+    /// the passed available_space and returns if this results in a concrete value (which it
+    /// will if the available_space is `Some`). Otherwise returns None.
```
```suggestion
    /// will if the available_space is `Some`). Otherwise returns `None`.
```

alice-i-cecile commented on src/style/grid.rs:
```diff
@@ -0,0 +1,446 @@
+    }
+}
+
+/// Represents the type of automatic reprepetition to perform
```
```suggestion
/// Represents the type of automatic repetition to perform
```

alice-i-cecile replied:
This could use a bit more explanation.

nicoburns commented:
> What type of review would be most useful to you?

I'm not sure. I realise that any kind of review is going to be difficult due to the size of the PR. There are a few areas that I think it would definitely be useful to get feedback on:

- The refactor of the `style` module (and addition of grid styles) + additional helpers in the `style_helpers` module
- The `compute::common::alignment` module. This module contains alignment logic which is common between Flexbox and Grid. Do we think that makes sense? Or should I try to keep them separate?
- The usage of the [grid](https://github.com/becheran/grid) crate (it provides basically a 2D Vec on top of an actual Vec). We could implement something similar as an internal module, but I figured it might make more sense to use an external implementation given that's it's pretty generic functionality. This crate isn't particularly widely used (I couldn't find one that was that didn't bring in a whole bunch of linear algebra math we don't need), but it's small enough that we could easily take on the maintenance if we needed to.

> My instincts are to merge this pretty aggressively, to... reduce the tooling pain that comes with massive PRs.

Yeah, I agree. I think correctness/robustness for this code will likely be achieved with a comprehensive test suite (which we might need to build up over time) rather than code review. And once we have this one merged we can follow up with much smaller PRs.

> to let users test it ASAP 

On this note, I figure it might be a good idea to do alpha `0.3` releases where we make it clear that the Grid implementation isn't ready for production yet. I think it'll take some time to squash all the bugs, and it would be nice if people could try it before it's 100% ready. Perhaps they'll even send us bug reports.

alice-i-cecile commented:
> The refactor of the style module (and addition of grid styles) + additional helpers in the style_helpers module

I like this!

> The compute::common::alignment module. This module contains alignment logic which is common between Flexbox and Grid. Do we think that makes sense? Or should I try to keep them separate?

This seems reasonable, but the exact details didn't feel quite right? I'd love if we could lump it in with other common functionality somewhere.

> The usage of the [grid](https://github.com/becheran/grid) crate (it provides basically a 2D Vec on top of an actual Vec). We could implement something similar as an internal module, but I figured it might make more sense to use an external implementation given that's it's pretty generic functionality. This crate isn't particularly widely used (I couldn't find one that was that didn't bring in a whole bunch of linear algebra math we don't need), but it's small enough that we could easily take on the maintenance if we needed to.

I'm kinda neutral on this. I'm fine with it for now.

> I figure it might be a good idea to do alpha 0.3 releases where we make it clear that the Grid implementation isn't ready for production yet

Yep I'd be happy with that!

nicoburns commented:
@alice-i-cecile

I believe I have addressed all of your concerns/requests (in particular I've filled in the release notes). I have also:

- Implemented absolute and hidden positioning
- Fixed support for children with margins (including absolutely positioned children).
- Added a whole bunch of tests and fixes. We're up to 100 gentests for grid now (compared to 277 for Flexbox). Still a lot more testing to be done, but I think this is a good start.
- Updated CI to run tests both with and without the grid feature flag. Unfortunately the only way I could get this to work was remove it from the "taffy" section in Cargo.toml, which means that my editor doesn't seem to pick up grid related files by default, `cargo test` doesn't include grid tests by default, and clippy doesn't lint grid files by default (you have to specify `--features experimental_grid`). Not sure if there is anything we can do about this. For now I've been adding it back it temporarily but not committing it. Perhaps we should make it a default features? Not sure.

As always, progress is being tracked over at https://github.com/DioxusLabs/taffy/issues/204. My suggestion is that we try to get this PR merged ASAP, but that we keep the issue open and continue to use it to track overall progress on Grid. I imagine the release notes might need some work, but I would like to suggest we start a new release issue/PR and don't block this PR on that.

alice-i-cecile reviewed (approved):
Awesome write-up! That was my last blocking request: I think that merging and iterating is the way to go now.
